### PR TITLE
SONAR-9324 add database migration purging developer data

### PR DESCRIPTION
**Pull request by sns-seb**

### CODE REVIEW

File: server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v65/PurgeDeveloperData.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+      case 0:
+      case 1:
+      case 2:
+        update.setString(1, uuid);
+        return true;
+      case 3:
+      case 4:
+        update.setLong(1, id);
+        return true;
+      case 5:
```
Comment (SonarTech) [inline server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v65/PurgeDeveloperData.java]: MAJOR This case's code block is the same as the block for the case on line 93. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1871)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by ehartmann** [comment]
Tested and looks good to me.

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchResponseLoader.java [hunk 2]
```diff
@@ -87,9 +97,107 @@ public SearchResponseData load(Collector collector, @Nullable Facets facets) {
...
+    result.addComponents(dbClient.componentDao().selectSubProjectsByComponentUuids(dbSession, collector.getComponentUuids()));
+    addProjectUuids(collector, dbSession, result);
+  }
+
+  private void addProjectUuids(Collector collector, DbSession dbSession, SearchResponseData result) {
+    Collection<ComponentDto> loadedComponents = result.getComponents();
+    for (ComponentDto component : loadedComponents) {
+      collector.addProjectUuid(component.projectUuid());
+    }
+    Set<String> loadedProjectUuids = loadedComponents.stream().filter(cpt -> cpt.uuid().equals(cpt.projectUuid())).map(ComponentDto::uuid).collect(MoreCollectors.toSet());
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchResponseLoader.java]: MAJOR A "NullPointerException" could be thrown; "loadedComponents" is nullable here. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchResponseLoader.java [hunk 3]
```diff
@@ -87,9 +97,107 @@ public SearchResponseData load(Collector collector, @Nullable Facets facets) {
...
+      } else {
+        List<UserDto> loadedUsers = dbClient.userDao().selectByLogins(dbSession, loginsToLoad);
+        result.setUsers(concat(preloadedUsers.stream(), loadedUsers.stream()).collect(toList(preloadedUsers.size() + loadedUsers.size())));
+      }
+    }
+  }
+
+  private void loadComponents(SearchResponseData preloadedResponseData, Collector collector, DbSession dbSession, SearchResponseData result) {
+    Collection<ComponentDto> preloadedComponents = preloadedResponseData.getComponents();
+    Set<String> preloadedComponentUuids = preloadedComponents.stream().map(ComponentDto::uuid).collect(MoreCollectors.toSet(preloadedComponents.size()));
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchResponseLoader.java]: MAJOR A "NullPointerException" could be thrown; "preloadedComponents" is nullable here. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: server/sonar-server/src/main/java/org/sonar/server/issue/IssueUpdater.java [hunk 4]
```diff
@@ -44,21 +49,46 @@ public IssueUpdater(DbClient dbClient, IssueStorage issueStorage, NotificationMa
...
+
+  public IssueDto saveIssue(DbSession session, DefaultIssue issue, IssueChangeContext context, @Nullable String comment) {
+    Optional<RuleDefinitionDto> rule = getRuleByKey(session, issue.getRuleKey());
+    ComponentDto project = dbClient.componentDao().selectOrFailByUuid(session, issue.projectUuid());
+    ComponentDto component = dbClient.componentDao().selectOrFailByUuid(session, issue.componentUuid());
+    return saveIssue(session, issue, context, comment, rule, project, component);
+  }
+
+  private IssueDto saveIssue(DbSession session, DefaultIssue issue, IssueChangeContext context, @Nullable String comment,
+    Optional<RuleDefinitionDto> rule, ComponentDto project, ComponentDto component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/IssueUpdater.java]: MINOR Specify a "RuleDefinitionDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**